File: mobile/apps/photos/lib/db/ml/schema.dart
import 'package:photos/services/machine_learning/face_ml/face_filtering/face_filtering_constants.dart';

const facesTable = 'faces';
const fileIDColumn = 'file_id';
const objectIdColumn = 'obj_id';
const faceIDColumn = 'face_id';
const faceDetectionColumn = 'detection';
const embeddingColumn = 'embedding';
const faceScore = 'score';
const faceBlur = 'blur';
const isSideways = 'is_sideways';
const imageWidth = 'width';
const imageHeight = 'height';
const mlVersionColumn = 'ml_version';
const personIdColumn = 'person_id';
const clusterIDColumn = 'cluster_id';
const personOrClusterIdColumn = 'person_or_cluster_id';
const textQueryColumn = 'text_query';
const createdAtColumn = 'created_at';

const createFacesTable =
    '''CREATE TABLE IF NOT EXISTS $facesTable (
  $fileIDColumn	INTEGER NOT NULL,
  $faceIDColumn  TEXT NOT NULL UNIQUE,
	$faceDetectionColumn	TEXT NOT NULL,
  $embeddingColumn BLOB NOT NULL,
  $faceScore  REAL NOT NULL,
  $faceBlur REAL NOT NULL DEFAULT $kLapacianDefault,
  $isSideways	INTEGER NOT NULL DEFAULT 0,
  $imageHeight	INTEGER NOT NULL DEFAULT 0,
  $imageWidth	INTEGER NOT NULL DEFAULT 0,
  $mlVersionColumn	INTEGER NOT NULL DEFAULT -1,
  PRIMARY KEY($fileIDColumn, $faceIDColumn)
  );
  ''';

const deleteFacesTable = 'DELETE FROM $facesTable';
const faceClustersTable = 'face_clusters';

const createFaceClustersTable =
    '''
CREATE TABLE IF NOT EXISTS $faceClustersTable (
  $faceIDColumn	TEXT NOT NULL,
  $clusterIDColumn TEXT NOT NULL,
  PRIMARY KEY($faceIDColumn)
);
''';
const fcClusterIDIndex =
    '''CREATE INDEX IF NOT EXISTS idx_fcClusterID ON $faceClustersTable($clusterIDColumn);''';
const deleteFaceClustersTable = 'DELETE FROM $faceClustersTable';
const clusterPersonTable = 'cluster_person';

const createClusterPersonTable =
    '''
CREATE TABLE IF NOT EXISTS $clusterPersonTable (
  $personIdColumn	TEXT NOT NULL,
  $clusterIDColumn	TEXT NOT NULL,
  PRIMARY KEY($personIdColumn, $clusterIDColumn)
);
''';
const deleteClusterPersonTable = 'DELETE FROM $clusterPersonTable';

const clusterSummaryTable = 'cluster_summary';
const avgColumn = 'avg';
const countColumn = 'count';
const createClusterSummaryTable =
    '''
CREATE TABLE IF NOT EXISTS $clusterSummaryTable (
  $clusterIDColumn	TEXT NOT NULL,
  $avgColumn BLOB NOT NULL,
  $countColumn INTEGER NOT NULL,
  PRIMARY KEY($clusterIDColumn)
);
''';

const deleteClusterSummaryTable = 'DELETE FROM $clusterSummaryTable';

const clusterCentroidVectorIdMappingTable = 'cluster_centroid_vector_id_map';
const clusterCentroidVectorIdColumn = 'cluster_vector_id';

const createClusterCentroidVectorIdMappingTable =
    '''
CREATE TABLE IF NOT EXISTS $clusterCentroidVectorIdMappingTable (
  $clusterCentroidVectorIdColumn INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
  $clusterIDColumn TEXT NOT NULL UNIQUE
);
''';

const deleteClusterCentroidVectorIdMappingTable =
    'DELETE FROM $clusterCentroidVectorIdMappingTable';

const notPersonFeedback = 'not_person_feedback';

const createNotPersonFeedbackTable =
    '''
CREATE TABLE IF NOT EXISTS $notPersonFeedback (
  $personIdColumn	TEXT NOT NULL,
  $clusterIDColumn TEXT NOT NULL,
  PRIMARY KEY($personIdColumn, $clusterIDColumn)
);
''';
const deleteNotPersonFeedbackTable = 'DELETE FROM $notPersonFeedback';
const clipTable = 'clip';

const createClipEmbeddingsTable =
    '''
CREATE TABLE IF NOT EXISTS $clipTable ( 
  $fileIDColumn INTEGER NOT NULL,
  $embeddingColumn BLOB NOT NULL,
  $mlVersionColumn INTEGER NOT NULL,
  PRIMARY KEY ($fileIDColumn)
  );
''';

const deleteClipEmbeddingsTable = 'DELETE FROM $clipTable';

const fileDataTable = 'filedata';
const createFileDataTable =
    '''
CREATE TABLE IF NOT EXISTS $fileDataTable ( 
  $fileIDColumn INTEGER NOT NULL,
  user_id INTEGER NOT NULL,
  type TEXT NOT NULL,
  size INTEGER NOT NULL,
  $objectIdColumn TEXT,
  obj_nonce TEXT,
  updated_at INTEGER NOT NULL,
  PRIMARY KEY ($fileIDColumn, type)
  );
''';

const deleteFileDataTable = 'DELETE FROM $fileDataTable';

const faceCacheTable = 'face_cache';

const createFaceCacheTable =
    '''
CREATE TABLE IF NOT EXISTS $faceCacheTable (
  $personOrClusterIdColumn TEXT NOT NULL UNIQUE,
  $faceIDColumn TEXT NOT NULL UNIQUE,
  PRIMARY KEY ($personOrClusterIdColumn)
);
''';

const deleteFaceCacheTable = 'DELETE FROM $faceCacheTable';

const textEmbeddingsCacheTable = 'text_embeddings_cache';

const createTextEmbeddingsCacheTable =
    '''
CREATE TABLE IF NOT EXISTS $textEmbeddingsCacheTable (
  $textQueryColumn TEXT NOT NULL,
  $embeddingColumn BLOB NOT NULL,
  $mlVersionColumn INTEGER NOT NULL,
  $createdAtColumn INTEGER NOT NULL,
  PRIMARY KEY ($textQueryColumn)
);
''';

const deleteTextEmbeddingsCacheTable = 'DELETE FROM $textEmbeddingsCacheTable';

const offlineFileKeyMapTable = 'offline_file_key_map';
const offlineFileKeyLocalIdColumn = 'local_id';
const offlineFileKeyIntIdColumn = 'local_int_id';

const createOfflineFileKeyMapTable =
    '''
CREATE TABLE IF NOT EXISTS $offlineFileKeyMapTable (
  $offlineFileKeyIntIdColumn INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
  $offlineFileKeyLocalIdColumn TEXT NOT NULL UNIQUE
);
''';

const petFaceIDColumn = 'pet_face_id';
const petBodyIDColumn = 'pet_body_id';
const speciesColumn = 'species';
const detectionColumn = 'detection';
const faceVectorIdColumn = 'face_vector_id';
const bodyVectorIdColumn = 'body_vector_id';

const petFacesTable = 'pet_faces';

const createPetFacesTable =
    '''CREATE TABLE IF NOT EXISTS $petFacesTable (
  $fileIDColumn INTEGER NOT NULL,
  $petFaceIDColumn TEXT NOT NULL UNIQUE,
  $faceDetectionColumn TEXT NOT NULL,
  $faceVectorIdColumn INTEGER UNIQUE,
  $speciesColumn INTEGER NOT NULL,
  $faceScore REAL NOT NULL,
  $imageHeight INTEGER NOT NULL DEFAULT 0,
  $imageWidth INTEGER NOT NULL DEFAULT 0,
  $mlVersionColumn INTEGER NOT NULL DEFAULT -1,
  PRIMARY KEY($fileIDColumn, $petFaceIDColumn)
);
''';

const deletePetFacesTable = 'DELETE FROM $petFacesTable';

const petBodiesTable = 'pet_bodies';
// Both tables use the SQL column "score".
const bodyScore = faceScore;

const createPetBodiesTable =
    '''CREATE TABLE IF NOT EXISTS $petBodiesTable (
  $fileIDColumn INTEGER NOT NULL,
  $petBodyIDColumn TEXT NOT NULL UNIQUE,
  $detectionColumn TEXT NOT NULL,
  $bodyVectorIdColumn INTEGER UNIQUE,
  $speciesColumn INTEGER NOT NULL,
  $bodyScore REAL NOT NULL,
  $imageHeight INTEGER NOT NULL DEFAULT 0,
  $imageWidth INTEGER NOT NULL DEFAULT 0,
  $mlVersionColumn INTEGER NOT NULL DEFAULT -1,
  PRIMARY KEY($fileIDColumn, $petBodyIDColumn)
);
''';

const deletePetBodiesTable = 'DELETE FROM $petBodiesTable';

const petFaceVectorIdMappingTable = 'pet_face_vector_id_map';
const petFaceVectorIdColumn = 'pet_face_vector_id';

const createPetFaceVectorIdMappingTable =
    '''
CREATE TABLE IF NOT EXISTS $petFaceVectorIdMappingTable (
  $petFaceVectorIdColumn INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
  $petFaceIDColumn TEXT NOT NULL UNIQUE
);
''';

const deletePetFaceVectorIdMappingTable =
    'DELETE FROM $petFaceVectorIdMappingTable';

const petBodyVectorIdMappingTable = 'pet_body_vector_id_map';
const petBodyVectorIdColumn = 'pet_body_vector_id';

const createPetBodyVectorIdMappingTable =
    '''
CREATE TABLE IF NOT EXISTS $petBodyVectorIdMappingTable (
  $petBodyVectorIdColumn INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
  $petBodyIDColumn TEXT NOT NULL UNIQUE
);
''';

const deletePetBodyVectorIdMappingTable =
    'DELETE FROM $petBodyVectorIdMappingTable';

File: rust/crates/photos/src/ml_db/schema.rs
pub const CREATE_FACES_TABLE: &str = "CREATE TABLE IF NOT EXISTS faces (
  file_id INTEGER NOT NULL,
  face_id TEXT NOT NULL UNIQUE,
  detection TEXT NOT NULL,
  embedding BLOB NOT NULL,
  score REAL NOT NULL,
  blur REAL NOT NULL DEFAULT 10000.0,
  is_sideways INTEGER NOT NULL DEFAULT 0,
  height INTEGER NOT NULL DEFAULT 0,
  width INTEGER NOT NULL DEFAULT 0,
  ml_version INTEGER NOT NULL DEFAULT -1,
  PRIMARY KEY(file_id, face_id)
);";

pub const CREATE_FACE_CLUSTERS_TABLE: &str = "CREATE TABLE IF NOT EXISTS face_clusters (
  face_id TEXT NOT NULL,
  cluster_id TEXT NOT NULL,
  PRIMARY KEY(face_id)
);";

pub const FC_CLUSTER_ID_INDEX: &str =
    "CREATE INDEX IF NOT EXISTS idx_fcClusterID ON face_clusters(cluster_id);";

pub const CREATE_CLUSTER_PERSON_TABLE: &str = "CREATE TABLE IF NOT EXISTS cluster_person (
  person_id TEXT NOT NULL,
  cluster_id TEXT NOT NULL,
  PRIMARY KEY(person_id, cluster_id)
);";

pub const CREATE_CLUSTER_SUMMARY_TABLE: &str = "CREATE TABLE IF NOT EXISTS cluster_summary (
  cluster_id TEXT NOT NULL,
  avg BLOB NOT NULL,
  count INTEGER NOT NULL,
  PRIMARY KEY(cluster_id)
);";

pub const CREATE_CLUSTER_CENTROID_VECTOR_ID_MAPPING_TABLE: &str =
    "CREATE TABLE IF NOT EXISTS cluster_centroid_vector_id_map (
  cluster_vector_id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
  cluster_id TEXT NOT NULL UNIQUE
);";

pub const CREATE_NOT_PERSON_FEEDBACK_TABLE: &str =
    "CREATE TABLE IF NOT EXISTS not_person_feedback (
  person_id TEXT NOT NULL,
  cluster_id TEXT NOT NULL,
  PRIMARY KEY(person_id, cluster_id)
);";

pub const CREATE_CLIP_EMBEDDINGS_TABLE: &str = "CREATE TABLE IF NOT EXISTS clip (
  file_id INTEGER NOT NULL,
  embedding BLOB NOT NULL,
  ml_version INTEGER NOT NULL,
  PRIMARY KEY (file_id)
);";

pub const CREATE_FILE_DATA_TABLE: &str = "CREATE TABLE IF NOT EXISTS filedata (
  file_id INTEGER NOT NULL,
  user_id INTEGER NOT NULL,
  type TEXT NOT NULL,
  size INTEGER NOT NULL,
  obj_id TEXT,
  obj_nonce TEXT,
  updated_at INTEGER NOT NULL,
  PRIMARY KEY (file_id, type)
);";

pub const CREATE_FACE_CACHE_TABLE: &str = "CREATE TABLE IF NOT EXISTS face_cache (
  person_or_cluster_id TEXT NOT NULL UNIQUE,
  face_id TEXT NOT NULL UNIQUE,
  PRIMARY KEY (person_or_cluster_id)
);";

pub const CREATE_TEXT_EMBEDDINGS_CACHE_TABLE: &str =
    "CREATE TABLE IF NOT EXISTS text_embeddings_cache (
  text_query TEXT NOT NULL,
  embedding BLOB NOT NULL,
  ml_version INTEGER NOT NULL,
  created_at INTEGER NOT NULL,
  PRIMARY KEY (text_query)
);";

pub const CREATE_PET_FACES_TABLE: &str = "CREATE TABLE IF NOT EXISTS pet_faces (
  file_id INTEGER NOT NULL,
  pet_face_id TEXT NOT NULL UNIQUE,
  detection TEXT NOT NULL,
  face_vector_id INTEGER UNIQUE,
  species INTEGER NOT NULL,
  score REAL NOT NULL,
  height INTEGER NOT NULL DEFAULT 0,
  width INTEGER NOT NULL DEFAULT 0,
  ml_version INTEGER NOT NULL DEFAULT -1,
  PRIMARY KEY(file_id, pet_face_id)
);";

pub const CREATE_PET_BODIES_TABLE: &str = "CREATE TABLE IF NOT EXISTS pet_bodies (
  file_id INTEGER NOT NULL,
  pet_body_id TEXT NOT NULL UNIQUE,
  detection TEXT NOT NULL,
  body_vector_id INTEGER UNIQUE,
  species INTEGER NOT NULL,
  score REAL NOT NULL,
  height INTEGER NOT NULL DEFAULT 0,
  width INTEGER NOT NULL DEFAULT 0,
  ml_version INTEGER NOT NULL DEFAULT -1,
  PRIMARY KEY(file_id, pet_body_id)
);";

pub const CREATE_PET_FACE_VECTOR_ID_MAPPING_TABLE: &str =
    "CREATE TABLE IF NOT EXISTS pet_face_vector_id_map (
  pet_face_vector_id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
  pet_face_id TEXT NOT NULL UNIQUE
);";

pub const CREATE_PET_BODY_VECTOR_ID_MAPPING_TABLE: &str =
    "CREATE TABLE IF NOT EXISTS pet_body_vector_id_map (
  pet_body_vector_id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
  pet_body_id TEXT NOT NULL UNIQUE
);";

pub const MIGRATION_SCRIPTS: [&str; 15] = [
    CREATE_FACES_TABLE,
    CREATE_FACE_CLUSTERS_TABLE,
    CREATE_CLUSTER_PERSON_TABLE,
    CREATE_CLUSTER_SUMMARY_TABLE,
    CREATE_NOT_PERSON_FEEDBACK_TABLE,
    FC_CLUSTER_ID_INDEX,
    CREATE_CLIP_EMBEDDINGS_TABLE,
    CREATE_FILE_DATA_TABLE,
    CREATE_FACE_CACHE_TABLE,
    CREATE_TEXT_EMBEDDINGS_CACHE_TABLE,
    CREATE_CLUSTER_CENTROID_VECTOR_ID_MAPPING_TABLE,
    CREATE_PET_FACES_TABLE,
    CREATE_PET_BODIES_TABLE,
    CREATE_PET_FACE_VECTOR_ID_MAPPING_TABLE,
    CREATE_PET_BODY_VECTOR_ID_MAPPING_TABLE,
];

pub const DELETE_FACES: &str = "DELETE FROM faces";
pub const DELETE_FACE_CLUSTERS: &str = "DELETE FROM face_clusters";
pub const DELETE_CLUSTER_PERSON: &str = "DELETE FROM cluster_person";
pub const DELETE_CLUSTER_SUMMARY: &str = "DELETE FROM cluster_summary";
pub const DELETE_CLUSTER_CENTROID_VECTOR_ID_MAPPING: &str =
    "DELETE FROM cluster_centroid_vector_id_map";
pub const DELETE_NOT_PERSON_FEEDBACK: &str = "DELETE FROM not_person_feedback";
pub const DELETE_CLIP_EMBEDDINGS: &str = "DELETE FROM clip";
pub const DELETE_FILE_DATA: &str = "DELETE FROM filedata";
pub const DELETE_PET_FACES: &str = "DELETE FROM pet_faces";
pub const DELETE_PET_BODIES: &str = "DELETE FROM pet_bodies";
pub const DELETE_PET_FACE_VECTOR_ID_MAPPING: &str = "DELETE FROM pet_face_vector_id_map";
pub const DELETE_PET_BODY_VECTOR_ID_MAPPING: &str = "DELETE FROM pet_body_vector_id_map";
